File: bookmark.py
from flask import Blueprint, abort, g, jsonify, render_template, request

from MyBookmarks.auth import login_required
from MyBookmarks.db import get_db

bp = Blueprint('bookmark', __name__)


@bp.route('/')
@login_required
def index():
    return render_template('base.html')


@bp.route('/bookmark')
@login_required
def bookmark():
    '''Show the bookmarks belong the current user.'''
    category_id = request.args.get('category')
    db = get_db()
    if category_id is None:
        category = {'id': -1, 'name': 'All Bookmarks'}
        bookmarks = db.execute('SELECT bookmark.id, bookmark, url, category FROM bookmark'
                               ' LEFT JOIN category ON category_id = category.id'
                               ' WHERE bookmark.user_id = ? ORDER BY seq',
                               (g.user['id'],)).fetchall()
        for i in bookmarks:
            if not i['category']:
                i['category'] = ''
    elif category_id == '0':
        category = {'id': 0, 'name': 'Uncategorized'}
        bookmarks = db.execute('SELECT id, bookmark, url FROM bookmark'
                               ' WHERE category_id = 0 AND user_id = ?'
                               ' ORDER BY seq', (g.user['id'],)).fetchall()
    else:
        category = {'id': int(category_id)}
        try:
            category['name'] = db.execute('SELECT category FROM category WHERE id = ? AND user_id = ?',
                                          (category_id, g.user['id'])).fetchone()['category']
        except TypeError:
            abort(403)
        bookmarks = db.execute('SELECT id, bookmark, url FROM bookmark'
                               ' WHERE category_id = ? AND user_id = ?'
                               ' ORDER BY seq', (category_id, g.user['id'])).fetchall()
        for i in bookmarks:
            i['category'] = category['name']
    return render_template('bookmark/index.html', category=category, bookmarks=bookmarks)


@bp.route('/category/get', methods=('GET',))
@login_required
def get_category():
    '''Get current user's categories.'''
    db = get_db()
    total = db.execute('SELECT count(bookmark) num FROM bookmark WHERE user_id = ?',
                       (g.user['id'],)).fetchone()['num']
    uncategorized = db.execute('SELECT count(bookmark) num FROM bookmark WHERE category_id = 0 AND user_id = ?',
                               (g.user['id'],)).fetchone()['num']
    categories = db.execute('SELECT category.id, category, count(bookmark) num'
                            ' FROM category LEFT JOIN bookmark ON category.id = category_id'
                            ' WHERE category.user_id = ? GROUP BY category.id ORDER BY category', (g.user['id'],)).fetchall()
    return jsonify({'total': total, 'uncategorized': uncategorized, 'categories': categories})


@bp.route('/category/add', methods=('GET', 'POST'))
@login_required
def add_category():
    '''Create a new category for the current user.'''
    if request.method == 'POST':
        db = get_db()
        category = request.form.get('category').strip()
        if category == '':
            message = 'Category name is empty.'
        elif len(category.encode('utf-8')) > 15:
            message = 'Category name exceeded length limit.'
        elif db.execute('SELECT id FROM category WHERE category = ? AND user_id = ?', (category, g.user['id'])).fetchone() is not None:
            message = f'Category {category} is already existed.'
        else:
            db.execute(
                'INSERT INTO category (category, user_id) VALUES (?, ?)', (category, g.user['id']))
            db.commit()
            return jsonify({'status': 1})
        return jsonify({'status': 0, 'message': message, 'error': 1})
    return render_template('bookmark/category.html', id=0, category={})


@bp.route('/category/edit/<int:id>', methods=('GET', 'POST'))
@login_required
def edit_category(id):
    '''Edit a category for the current user.'''
    db = get_db()
    category = db.execute(
        'SELECT * FROM category WHERE id = ? AND user_id = ?', (id, g.user['id'])).fetchone()
    if not category:
        abort(403)
    if request.method == 'POST':
        old = category['category']
        new = request.form.get('category').strip()
        error = 0
        if new == '':
            message = 'New category name is empty.'
            error = 1
        elif old == new:
            message = 'New category is same as old category.'
        elif len(new.encode('utf-8')) > 15:
            message = 'Category name exceeded length limit.'
            error = 1
        elif db.execute('SELECT id FROM category WHERE category = ? AND user_id = ?', (new, g.user['id'])).fetchone() is not None:
            message = f'Category {new} is already existed.'
            error = 1
        else:
            db.execute(
                'UPDATE category SET category = ? WHERE id = ? AND user_id = ?', (new, id, g.user['id']))
            db.commit()
            return jsonify({'status': 1})
        return jsonify({'status': 0, 'message': message, 'error': error})
    return render_template('bookmark/category.html', id=id, category=category)


@bp.route('/category/delete/<int:id>', methods=('POST',))
@login_required
def delete_category(id):
    '''Edit a category for the current user.'''
    db = get_db()
    db.execute('DELETE FROM category WHERE id = ? and user_id = ?',
               (id, g.user['id']))
    db.execute('UPDATE bookmark SET category_id = 0 WHERE category_id = ? and user_id = ?',
               (id, g.user['id']))
    db.commit()
    return jsonify({'status': 1})


def get_category_id(category, user_id):
    if category:
        db = get_db()
        category_id = db.execute(
            'SELECT id FROM category WHERE category = ? AND user_id = ?', (category, user_id)).fetchone()
        if len(category.encode('utf-8')) > 15:
            return None
        elif category_id:
            return category_id['id']
        else:
            db.execute(
                'INSERT INTO category (category, user_id) VALUES (?, ?)', (category, user_id))
            return db.execute('SELECT last_insert_rowid() id').fetchone()['id']
    else:
        return 0


@bp.route('/bookmark/add', methods=('GET', 'POST'))
@login_required
def add_bookmark():
    '''Create a new bookmark for the current user.'''
    category_id = request.args.get('category_id')
    db = get_db()
    if category_id:
        category = db.execute('SELECT category FROM category WHERE id = ? AND user_id = ?',
                              (category_id, g.user['id'])).fetchone()['category']
    else:
        category = ''
    categories = db.execute(
        'SELECT category FROM category WHERE user_id = ? ORDER BY category', (g.user['id'],)).fetchall()
    if request.method == 'POST':
        category = request.form.get('category').strip()
        bookmark = request.form.get('bookmark').strip()
        url = request.form.get('url').strip()
        category_id = get_category_id(category, g.user['id'])
        error = 0
        if bookmark == '':
            message = f'Bookmark name is empty.'
            error = 1
        elif db.execute('SELECT id FROM bookmark WHERE bookmark = ? AND user_id = ?', (bookmark, g.user['id'])).fetchone() is not None:
            message = f'Bookmark name {bookmark} is already existed.'
            error = 1
        elif db.execute('SELECT id FROM bookmark WHERE url = ? AND user_id = ?', (url, g.user['id'])).fetchone() is not None:
            message = f'Bookmark url {url} is already existed.'
            error = 2
        elif category_id is None:
            message = 'Category name exceeded length limit.'
            error = 3
        else:
            db.execute('INSERT INTO bookmark (bookmark, url, user_id, category_id)'
                       ' VALUES (?, ?, ?, ?)', (bookmark, url, g.user['id'], category_id))
            db.commit()
            return jsonify({'status': 1})
        return jsonify({'status': 0, 'message': message, 'error': error})
    return render_template('bookmark/bookmark.html', id=0, bookmark={'category': category}, categories=categories)


@bp.route('/bookmark/edit/<int:id>', methods=('GET', 'POST'))
@login_required
def edit_bookmark(id):
    '''Edit a bookmark for the current user.'''
    db = get_db()
    bookmark = db.execute('SELECT bookmark, url, category FROM bookmark'
                          ' LEFT JOIN category ON category_id = category.id'
                          ' WHERE bookmark.id = ? AND bookmark.user_id = ?',
                          (id, g.user['id'])).fetchone()
    if not bookmark:
        abort(403)
    else:
        if not bookmark['category']:
            bookmark['category'] = ''
    categories = db.execute(
        'SELECT category FROM category WHERE user_id = ? ORDER BY category', (g.user['id'],)).fetchall()
    if request.method == 'POST':
        old = (bookmark['bookmark'], bookmark['url'], bookmark['category'])
        bookmark = request.form.get('bookmark').strip()
        url = request.form.get('url').strip()
        category = request.form.get('category').strip()
        category_id = get_category_id(category, g.user['id'])
        error = 0
        if bookmark == '':
            message = f'Bookmark name is empty.'
            error = 1
        elif old == (bookmark, url, category):
            message = 'New bookmark is same as old bookmark.'
        elif db.execute('SELECT id FROM bookmark WHERE bookmark = ? AND id != ? AND user_id = ?', (bookmark, id, g.user['id'])).fetchone() is not None:
            message = f'Bookmark name {bookmark} is already existed.'
            error = 1
        elif db.execute('SELECT id FROM bookmark WHERE url = ? AND id != ? AND user_id = ?', (url, id, g.user['id'])).fetchone() is not None:
            message = f'Bookmark url {url} is already existed.'
            error = 2
        elif category_id is None:
            message = 'Category name exceeded length limit.'
            error = 3
        else:
            db.execute('UPDATE bookmark SET bookmark = ?, url = ?, category_id = ?'
                       ' WHERE id = ? AND user_id = ?', (bookmark, url, category_id, id, g.user['id']))
            db.commit()
            return jsonify({'status': 1})
        return jsonify({'status': 0, 'message': message, 'error': error})
    return render_template('bookmark/bookmark.html', id=id, bookmark=bookmark, categories=categories)


@bp.route('/bookmark/delete/<int:id>', methods=('POST',))
@login_required
def delete_bookmark(id):
    '''Edit a bookmark for the current user.'''
    db = get_db()
    db.execute('DELETE FROM bookmark WHERE id = ? and user_id = ?',
               (id, g.user['id']))
    db.commit()
    return jsonify({'status': 1})


@bp.route('/reorder', methods=('POST',))
@login_required
def reorder():
    orig = request.form.get('orig')
    dest = request.form.get('dest')
    refer = request.form.get('refer')
    db = get_db()
    orig_seq = db.execute(
        'SELECT seq FROM bookmark WHERE bookmark = ? AND user_id = ?', (orig, g.user['id'])).fetchone()['seq']
    if dest != '#TOP_POSITION#':
        dest_seq = db.execute(
            'SELECT seq FROM bookmark WHERE bookmark = ? AND user_id = ?', (dest, g.user['id'])).fetchone()['seq']
    else:
        dest_seq = db.execute(
            'SELECT seq FROM bookmark WHERE bookmark = ? AND user_id = ?', (refer, g.user['id'])).fetchone()['seq']-1
    if orig_seq > dest_seq:
        dest_seq += 1
        db.execute('UPDATE bookmark SET seq = seq+1 WHERE seq >= ? AND user_id = ? AND seq < ?',
                   (dest_seq, g.user['id'], orig_seq))
    else:
        db.execute('UPDATE bookmark SET seq = seq-1 WHERE seq <= ? AND user_id = ? AND seq > ?',
                   (dest_seq, g.user['id'], orig_seq))
    db.execute('UPDATE bookmark SET seq = ? WHERE bookmark = ? AND user_id = ?',
               (dest_seq, orig, g.user['id']))
    db.commit()
    return '1'

File: bookmark.go
package main

import (
	"fmt"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/sunshineplan/database/mongodb"
)

type bookmark struct {
	ID       string `json:"id"`
	ObjectID string `json:"_id,omitempty" bson:"_id"`
	Bookmark string `json:"bookmark"`
	URL      string `json:"url"`
	Category string `json:"category"`
	Seq      int    `json:"seq"`
}

func getBookmark(c *gin.Context) {
	var r struct{ Start int64 }
	c.ShouldBindJSON(&r)
	userID, _ := c.Get("id")
	bookmarks := []*bookmark{}
	if err := bookmarkClient.Find(
		mongodb.M{"user": userID}, &mongodb.FindOpt{Sort: mongodb.M{"seq": 1}, Skip: r.Start, Limit: 50}, &bookmarks,
	); err != nil {
		svc.Println("Failed to query bookmarks:", err)
		c.AbortWithStatus(500)
		return
	}
	for _, i := range bookmarks {
		i.ID = i.ObjectID
		i.ObjectID = ""
	}
	c.JSON(200, bookmarks)
}

func addBookmark(c *gin.Context) {
	var data bookmark
	if err := c.BindJSON(&data); err != nil {
		svc.Print(err)
		return
	}

	if data.Bookmark == "" {
		c.JSON(200, gin.H{"status": 0, "message": "Bookmark name is empty.", "error": 1})
		return
	} else if len(data.Category) > 15 {
		c.JSON(200, gin.H{"status": 0, "message": "Category name exceeded length limit.", "error": 3})
		return
	} else if data.Category == "All Bookmarks" || data.Category == "Uncategorized" {
		c.JSON(200, gin.H{"status": 0, "message": "Category name is not allow.", "error": 3})
		return
	}

	userID, _ := c.Get("id")
	ec := make(chan error, 2)
	var exist1, exist2 bool
	go func() {
		var err error
		exist1, err = checkExist(mongodb.M{"bookmark": data.Bookmark, "user": userID})
		ec <- err
	}()
	go func() {
		var err error
		exist2, err = checkExist(mongodb.M{"url": data.URL, "user": userID})
		ec <- err
	}()
	for i := 0; i < 2; i++ {
		if err := <-ec; err != nil {
			svc.Println("Failed to get bookmark:", err)
			c.AbortWithStatus(500)
			return
		}
	}

	var message string
	var errorCode int
	switch {
	case exist1:
		message = fmt.Sprintf("Bookmark name %s is already existed.", data.Bookmark)
		errorCode = 1
	case exist2:
		message = fmt.Sprintf("Bookmark url %s is already existed.", data.URL)
		errorCode = 2
	default:
		var bookmarks []bookmark
		if err := bookmarkClient.Find(
			mongodb.M{"user": userID}, &mongodb.FindOpt{Sort: mongodb.M{"seq": -1}, Limit: 1}, &bookmarks,
		); err != nil {
			svc.Println("Failed to get bookmarks:", err)
			c.AbortWithStatus(500)
			return
		}

		var seq int
		if len(bookmarks) == 0 {
			seq = 1
		} else {
			seq = bookmarks[0].Seq + 1
		}

		doc := struct {
			Bookmark string `json:"bookmark" bson:"bookmark"`
			URL      string `json:"url" bson:"url"`
			User     any    `json:"user" bson:"user"`
			Seq      int    `json:"seq" bson:"seq"`
			Created  any    `json:"created" bson:"created"`
			Category string `json:"category,omitempty" bson:"category,omitempty"`
		}{
			Bookmark: data.Bookmark,
			URL:      data.URL,
			User:     userID,
			Seq:      seq,
			Created:  bookmarkClient.Date(time.Now()).Interface(),
			Category: data.Category,
		}

		insertedID, err := bookmarkClient.InsertOne(doc)
		if err != nil {
			svc.Println("Failed to add bookmark:", err)
			c.AbortWithStatus(500)
			return
		}

		newLastModified(userID, c)
		c.JSON(200, gin.H{"status": 1, "id": insertedID.(mongodb.ObjectID).Hex(), "seq": seq})
		return
	}

	c.JSON(200, gin.H{"status": 0, "message": message, "error": errorCode})
}

func editBookmark(c *gin.Context) {
	id, err := bookmarkClient.ObjectID(c.Param("id"))
	if err != nil {
		svc.Print(err)
		c.AbortWithStatus(400)
		return
	}

	var new bookmark
	if err := c.BindJSON(&new); err != nil {
		svc.Print(err)
		return
	}

	if new.Bookmark == "" {
		c.JSON(200, gin.H{"status": 0, "message": "Bookmark name is empty.", "error": 1})
		return
	} else if len(new.Category) > 15 {
		c.JSON(200, gin.H{"status": 0, "message": "Category name exceeded length limit.", "error": 3})
		return
	} else if new.Category == "All Bookmarks" || new.Category == "Uncategorized" {
		c.JSON(200, gin.H{"status": 0, "message": "Category name is not allow.", "error": 3})
		return
	}

	userID, _ := c.Get("id")
	ec := make(chan error, 3)
	var old bookmark
	var exist1, exist2 bool
	go func() {
		ec <- bookmarkClient.FindOne(mongodb.M{"_id": id.Interface(), "user": userID}, nil, &old)
	}()
	go func() {
		var err error
		exist1, err = checkExist(mongodb.M{"_id": mongodb.M{"$ne": id.Interface()}, "bookmark": new.Bookmark, "user": userID})
		ec <- err
	}()
	go func() {
		var err error
		exist2, err = checkExist(mongodb.M{"_id": mongodb.M{"$ne": id.Interface()}, "url": new.URL, "user": userID})
		ec <- err
	}()
	for i := 0; i < 3; i++ {
		if err := <-ec; err != nil {
			svc.Println("Failed to get bookmark:", err)
			c.AbortWithStatus(500)
			return
		}
	}

	var message string
	var errorCode int
	switch {
	case old == new:
		message = "New bookmark is same as old bookmark."
	case exist1:
		message = fmt.Sprintf("Bookmark name %s is already existed.", new.Bookmark)
		errorCode = 1
	case exist2:
		message = fmt.Sprintf("Bookmark url %s is already existed.", new.URL)
		errorCode = 2
	default:
		var update mongodb.M
		if new.Category == "" {
			update = mongodb.M{"$set": mongodb.M{"bookmark": new.Bookmark, "url": new.URL}, "$unset": mongodb.M{"category": ""}}
		} else {
			update = mongodb.M{"$set": mongodb.M{"bookmark": new.Bookmark, "url": new.URL, "category": new.Category}}
		}
		if _, err := bookmarkClient.UpdateOne(mongodb.M{"_id": id.Interface()}, update, nil); err != nil {
			svc.Println("Failed to edit bookmark:", err)
			c.AbortWithStatus(500)
			return
		}

		newLastModified(userID, c)
		c.JSON(200, gin.H{"status": 1})
		return
	}

	c.JSON(200, gin.H{"status": 0, "message": message, "error": errorCode})
}

func deleteBookmark(c *gin.Context) {
	id, err := bookmarkClient.ObjectID(c.Param("id"))
	if err != nil {
		svc.Print(err)
		c.AbortWithStatus(400)
		return
	}
	userID, _ := c.Get("id")
	var bookmark bookmark
	if err := bookmarkClient.FindOneAndDelete(mongodb.M{"_id": id.Interface()}, nil, &bookmark); err != nil {
		if err == mongodb.ErrNoDocuments {
			c.AbortWithStatus(403)
			return
		}
		svc.Println("Failed to delete bookmark:", err)
		c.AbortWithStatus(500)
		return
	}

	if _, err := bookmarkClient.UpdateMany(
		mongodb.M{"user": userID, "seq": mongodb.M{"$gt": bookmark.Seq}},
		mongodb.M{"$inc": mongodb.M{"seq": -1}},
		nil,
	); err != nil {
		svc.Println("Failed to reorder after delete bookmark:", err)
		c.AbortWithStatus(500)
		return
	}

	newLastModified(userID, c)
	c.JSON(200, gin.H{"status": 1})
}

func reorder(c *gin.Context) {
	var data struct{ Orig, Dest string }
	if err := c.BindJSON(&data); err != nil {
		svc.Print(err)
		return
	}

	origID, err := bookmarkClient.ObjectID(data.Orig)
	if err != nil {
		svc.Print(err)
		c.AbortWithStatus(400)
		return
	}
	destID, err := bookmarkClient.ObjectID(data.Dest)
	if err != nil {
		svc.Print(err)
		c.AbortWithStatus(400)
		return
	}

	userID, _ := c.Get("id")
	for _, id := range []mongodb.ObjectID{origID, destID} {
		if exist, err := checkExist(mongodb.M{"_id": id.Interface(), "user": userID}); err != nil {
			svc.Println("Failed to get bookmark:", err)
			c.AbortWithStatus(500)
			return
		} else if !exist {
			c.AbortWithStatus(403)
			return
		}
	}

	if err := reorderBookmark(userID, origID, destID); err != nil {
		svc.Println("Failed to reorder bookmark:", err)
		c.AbortWithStatus(500)
		return
	}
	c.String(200, "1")
}
